packages/next/client/components/reducer.ts

```diff
@@ -330,67 +330,118 @@ export const ACTION_NAVIGATE = 'navigate'
+ *      - Applies the new router state.
+ *   - optimistic
+ *      - Enabled when somewhere in the router state path to the page there is a loading.js.
+ *      - Similar to `soft` but kicks off the data fetch in the reducer and applies `data` in the spot that should suspend.
+ *      - This enables showing loading states while navigating.
+ *      - Will trigger fast path or server-patch case in layout-router.
+ * - `soft`
+ *   - Reuses the existing cache.
+ *   - Creates an optimistic router state that causes the fetch to start in layout-router when there is missing data.
+ *   - If there is no missing data the existing cache data is rendered.
+ */
+interface NavigateAction {
+  type: typeof ACTION_NAVIGATE
+  url: URL
+  cacheType: 'soft' | 'hard'
+  navigateType: 'push' | 'replace'
+  cache: CacheNode
+  mutable: {
+    previousTree?: FlightRouterState
+    patchedTree?: FlightRouterState
+  }
+}
+
+/**
+ * Restore applies the provided router state.
+ * - Only used for `popstate` (back/forward navigation) where a known router state has to be applied.
+ * - Router state is applied as-is from the history state.
+ * - If any data is missing it will be fetched in layout-router during rendering and trigger fast path or server-patch case.
+ * - If no data is missing the existing cached data is rendered.
+ */
+interface RestoreAction {
+  type: typeof ACTION_RESTORE
+  url: URL
+  tree: FlightRouterState
+}
+
+/**
+ * Server-patch applies the provided Flight data to the cache and router tree.
+ * - Only triggered in layout-router when the data can't be handled in the fast path.
+ * - Main case where this is triggered is when a rewrite applies and Flight data for a different path is returned from the server.
+ * - Creates a new cache and router state with the Flight data applied.
+ */
+interface ServerPatchAction {
+  type: typeof ACTION_SERVER_PATCH
+  flightData: FlightData
+  previousTree: FlightRouterState
+  cache: CacheNode
+}
+
+/**
+ * Reducer that handles the app-router state updates.
+ */
 export function reducer(
   state: AppRouterState,
-  action:
-    | {
-        type: typeof ACTION_RELOAD
-        url: URL
-        cache: CacheNode
-        mutable: {
-          previousTree?: FlightRouterState
-          patchedTree?: FlightRouterState
-        }
-      }
-    | {
-        type: typeof ACTION_NAVIGATE
-
-        url: URL
-        cacheType: 'soft' | 'hard'
-        navigateType: 'push' | 'replace'
-        cache: CacheNode
-        mutable: {
-          previousTree?: FlightRouterState
-          patchedTree?: FlightRouterState
-        }
-      }
-    | {
-        type: typeof ACTION_RESTORE
-        url: URL
-        tree: FlightRouterState
-      }
-    | {
-        type: typeof ACTION_SERVER_PATCH
-        flightData: FlightData
-        previousTree: FlightRouterState
-        cache: CacheNode
-      }
+  action: ReloadAction | NavigateAction | RestoreAction | ServerPatchAction
```

Comment:
Much cleaner 👍🏻

Reply:
Also might be valuable to mark these (`state` and `action`) properties as `Readonly`?

```ts
export function reducer(
  state: Readonly<AppRouterState>,
  action: Readonly<
    ReloadAction | NavigateAction | RestoreAction | ServerPatchAction
  >
): AppRouterState;
```

Enforcing this on the type level might be helpful to prevent mistakes in the future!

Reply:
Makes sense! Seems it prevents at least one form of misusing it, notably the `action` is mutated just at a lower level (`mutable`) in some actions.